src/ir_passes/lower/fsm.rs

```diff
@@ -1,17 +1,540 @@
+                    let done = prim std_reg(1);
+                    let final_state = constant(states-1, bitwidth);
+                );
+
+                // This component's interface
+                let this = builder.component.signature.clone();
+
+                // checks if the counter is currently on the final state.
+                // state == _{n-1}
+                let rst_check =
+                    guard!(state["out"]).eq(guard!(final_state["out"]));
+
+                // check if we should increment the counter.
+                // go || state != 0
+                let go_check = guard!(this["go"])
+                    .or(guard!(state["out"]).eq(guard!(zero["out"])).not());
+
+                let enable_check = rst_check.clone().or(go_check.clone());
+
+                // go && state == 0
+                let zero_check = guard!(this["go"])
+                    .and(guard!(state["out"]).eq(guard!(zero["out"])));
+
+                // add base assignments
+                builder.component.continuous_assignments.extend(
+                    build_assignments!(builder;
+                        // build assignments for state+1
+                        add["left"] = ? state["out"];
+                        add["right"] = ? one["out"];
+                        // hook up adder to register
+                        state["in"] = go_check ? add["out"];
+                        // hook up zero when counter finished
+                        state["in"] = rst_check ? zero["out"];
+                        // always enable register when there is an input
+                        state["write_en"] = enable_check ? signal_on["out"];
+                        // _0 = go && state == 0
+                        this["_0"] = zero_check ? signal_on["out"];
+                        // hook up state
+                        this["state"] = ? state["out"];
+                        // done <= _{n-1};
+                        // unused at the moment but useful if we want to chain FSMs.
+                        this["done"] = ? done["out"];
+                        done["in"] = rst_check ? signal_on["out"];
+                        done["write_en"] = ? signal_on["out"];
+                    ),
+                );
+
+                drop(this);
+                comp
+            })
+    }
+
+    /// Helper function that generates a [calyx::Component] for an fsm with `n` states.
+    fn add_simple(
+        &mut self,
+        states: u64,
+        lib: Option<&calyx::LibrarySignatures>,
+    ) -> &calyx::Component {
+        self.fsms.entry(FsmType::Simple(states)).or_insert_with(|| {
+            let ports: Vec<calyx::PortDef<u64>> = (0..states)
+                // create the state ports in the format `_state`.
+                .map(|n| {
+                    (
+                        calyx::Id::from(format!("_{n}")),
+                        1,
+                        calyx::Direction::Output,
+                    )
+                        .into()
```

Comment:
PortDef constructions in Calyx is so ugly. We should probably create a macro for this in `calyx_ir`. Maybe worth opening an issue in the Calyx repo if you agree.

Reply:
Done! https://github.com/cucapra/calyx/issues/1668.